File: src/main/java/com/github/unchama/seichiassist/commands/GiganticFeverCommand.kt
package com.github.unchama.seichiassist.commands

import com.github.unchama.contextualexecutor.asNonBlockingTabExecutor
import com.github.unchama.contextualexecutor.builder.ContextualExecutorBuilder
import com.github.unchama.seichiassist.ManagedWorld
import com.github.unchama.seichiassist.SeichiAssist
import com.github.unchama.seichiassist.util.Util
import com.github.unchama.targetedeffect.EmptyEffect
import kotlinx.coroutines.delay
import org.bukkit.ChatColor
import org.bukkit.Difficulty

object GiganticFeverCommand {
  private val worldsToToggleDifficulty = ManagedWorld.seichiWorlds.map { it.name }

  val executor = ContextualExecutorBuilder.beginConfiguration()
      .execution {
        val config = SeichiAssist.seichiAssistConfig

        Util.sendEveryMessage("${ChatColor.AQUA}フィーバー！この時間MOBたちは踊りに出かけてるぞ！今が整地時だ！")
        Util.sendEveryMessage("${ChatColor.AQUA}(${config.giganticFeverDisplayTime}間)")

        Util.setDifficulty(worldsToToggleDifficulty, Difficulty.PEACEFUL)

        delay(config.giganticFeverMinutes * 60L * 1000L)

        Util.setDifficulty(worldsToToggleDifficulty, Difficulty.HARD)
        Util.sendEveryMessage("${ChatColor.AQUA}フィーバー終了！MOBたちは戻ってきたぞ！")

        EmptyEffect
      }
      .build()
      .asNonBlockingTabExecutor()
}

File: src/main/scala/com/github/unchama/seichiassist/commands/GiganticFeverCommand.scala
package com.github.unchama.seichiassist.commands

import cats.effect.IO
import com.github.unchama.contextualexecutor.builder.ContextualExecutorBuilder
import com.github.unchama.seichiassist.concurrent.PluginExecutionContexts.onMainThread
import com.github.unchama.seichiassist.util.{SendMessageEffect, WorldSettings}
import com.github.unchama.seichiassist.{ManagedWorld, SeichiAssist}
import org.bukkit.ChatColor._
import org.bukkit.Difficulty
import org.bukkit.command.TabExecutor

import scala.concurrent.ExecutionContext

object GiganticFeverCommand {
  import cats.implicits._
  import scala.concurrent.duration._

  private val worldsToToggleDifficulty = ManagedWorld.seichiWorlds.map(_.alphabetName).toList

  val executor: TabExecutor = ContextualExecutorBuilder
    .beginConfiguration
    .buildWithIOAsExecution {
      val config = SeichiAssist.seichiAssistConfig
      val broadcast = SendMessageEffect.sendMessageToEveryoneIgnoringPreferenceIO(_: String)

      List(
        broadcast(s"${AQUA}フィーバー！この時間MOBたちは踊りに出かけてるぞ！今が整地時だ！"),
        broadcast(s"$AQUA(${config.getGiganticFeverDisplayTime}間)"),
        WorldSettings.setDifficulty(worldsToToggleDifficulty, Difficulty.PEACEFUL),
        IO.sleep((config.getGiganticFeverMinutes * 60).minutes)(
          IO.timer(ExecutionContext.global)
        ),
        WorldSettings.setDifficulty(worldsToToggleDifficulty, Difficulty.HARD),
        broadcast(s"${AQUA}フィーバー終了！MOBたちは戻ってきたぞ！")
      ).sequence.void
    }
    .asNonBlockingTabExecutor()
}
